src/libraries/Microsoft.Extensions.Logging/src/LoggerFactoryScopeProvider.cs

```diff
@@ -122,17 +122,17 @@ public void Dispose()
             {
                 if (!_isDisposed)
                 {
-                    _provider._currentScope.Value = Parent;
+                    _provider._currentScope.Value = Parent!;
```

Comment:
Wouldn't it be better to annotate:

```C#
private readonly AsyncLocal<Scope?> _currentScope = new AsyncLocal<Scope?>();
```

We do set it to `null` in places. So it seems to be the correct annotation.

Reply:
I was thinking about it, but objects are allowed to have invalid state after being disposed, so it's ok to possibly set non-nullable field to `null` inside `Dispose()`.
It's not set to `null` in other places as far as I remember.

Reply:
> objects are allowed to have invalid state after being disposed

But the state being invalid here is not on the object being disposed. The object being disposed is the `Scope` object. So if the last one is disposed, `_currentScope.Value` is going to be `null`.

We already assume `_currentScope.Value` could be `null` everywhere we access it. So it seems to make more sense to just encode it in the Type as `AsyncLocal<Scope?>`.